### ISSUE TITLE
InfluxDB demo hub container panics after a few messages

### ISSUE THREAD

**Issue by TheYkk** [build-trust/ockam#981]
I was trying example usage of ockam by [here](https://www.ockam.io/learn/how-to-guides/using-add-ons/enterprise/influxdb/connect-and-use-ockam-hub#step-1-clone-the-ockam-repo-to-get-the-demo-scripts)

And I pushed data with `./tools/docker/demo/influxdb.sh telegraf-write` but after few try influxdb and ockam stopped
Version : v0.10.1

## Error logs

Ockam Hub logs:
```
Router public key: e0d41f8190269576c8cc2e723a741c0b00aa8ec41904e5f35f216e507a7b1257
Channel cleartext address: 4b7f0bd6
thread 'main' panicked at 'range start index 1 out of range for slice of length 0', message/src/lib.rs:83:26
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

Influxdb Cckamd  logs:
```
Responder public key: ad4c43cf6efab2abb2055868aba0d0faae4804dda5c640bf724e4f25b9103969
Service address: 01242020
Channel cleartext address: afe9d0f9
Channel cleartext address: 09962e9f
thread 'main' panicked at 'channel manager poll failure: ChannelError { inner: 
The requested feature is not supported }', daemon/src/node.rs:200:30
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

**Comment by mrinalwadhwa** [comment]
@TheYkk thank you for reporting the error 🙏

**Comment by gabhijit** [comment]
@mrinalwadhwa : I am able to reproduce this as well. Also, it might be noted that this error occurs if you send messages in a burst. For example if I do something like - 

```
for i in `seq 1 10`; do sudo ./tools/docker/demo/influxdb.sh telegraf-write; done
```
this causes the `ifnluxdb-ockamd` container to crash. 

However if one does -(Note: `sleep 1` after every `telegraf-write`) 

```
for i in `seq 1 10`; do sudo ./tools/docker/demo/influxdb.sh telegraf-write; sleep 1; done
```
This works fine. 

So it's more than the number of messages burst that causes the problem.

**Comment by gabhijit** [comment]
I tried to look at the problem in a bit more detail over the weekend and here is what is actually happening. 

1. Since TCP is a stream based protocol, reads are not guaranteed to be on (application level) message boundaries. When you send a burst of messages, data corresponding to  many messages is sent and read together. This does not happen if you send messages one by one because a message is contained in a single TCP segment and often read entirely in a single `read` (very often but not always).  Since the messages do not explicitly encode the length of the payload (or the message), entire read message is seen as a single message, which cannot be decrypted and that causes the panic. 

2. I built the containers again with some debug prints and can be seen below that this is what happens.. 
```rust
m.messagetype: Payload, m.message_body_length: 148
m.messagetype: Payload, m.message_body_length: 106
Before Handle Receive:
Handle Receive: Payload
Before Handle Receive:
Handle Receive: Payload
Before Handle Receive:
m.messagetype: Payload, m.message_body_length: 840 <------ This causes trouble
Handle Receive: Payload
VaultFailErrorKind: AeadAesGcm
thread 'main' panicked at 'channel manager poll failure: ChannelError { inner:    0: failure::backtrace::internal::InternalBacktrace::new
   1: <failure::backtrace::Backtrace as core::default::Default>::default
   2: <ockam_channel::error::ChannelError as core::convert::From<ockam_vault::error::VaultFailError>>::from
   3: ockam_channel::ChannelManager<I,R,E>::handle_recv
   4: ockam_channel::ChannelManager<I,R,E>::poll
   5: ockamd::node::Node::run
   6: ockamd::main
   7: std::sys_common::backtrace::__rust_begin_short_backtrace
```
Unfortunately there isn't an easy way to fix this without actually changing the message structures. Individual message types need to carry the length encoded so that only that part of the message is decoded and remaining part is not decoded immediately. Also, this has implications for the APIs, which is somewhat involved to fix immediately.

2. The crashing `ockamd` in the `influxdb-ockamd` container causes it to send an RST to the `ockam-hub` which results in an EOF and hence 0 bytes read. In the current implementation handling 0 bytes read as a special case of connection closed is not handled, and this causes the panic in the decode (trying to decode what is not received). This part is fairly easy to be fixed by handling this correctly in the current implementation so that at-least the `ockam-hub` does not crash on any of the connected peers sending an RST. This is the first error above in the `ockam-hub` container.

3. Also not reported above - but happens `telegraf-ockamd` container goes in a loop trying to connect to `ockam-hub` after it is crashed. This outputs a bunch of logs when that happens. There might be a back-off added to this, to avoid too many logs being generated. 

4. Some minor improvements to the shell script like passing a `--debug` parameter so that the environment `RUST_BACKTRACE=1` can be set and passed to the container command line to get more information when any errors occur might also be done.
